src/hotspot/share/opto/inlinetypenode.cpp

```diff
@@ -1009,34 +1009,22 @@ InlineTypeNode* InlineTypeNode::make_from_multi(GraphKit* kit, MultiNode* multi,
   return kit->gvn().transform(vt)->as_InlineType();
 }
 
-InlineTypeNode* InlineTypeNode::make_larval(GraphKit* kit, bool allocate) const {
+Node* InlineTypeNode::make_larval(GraphKit* kit) const {
   ciInlineKlass* vk = inline_klass();
-  InlineTypeNode* res = make_uninitialized(kit->gvn(), vk);
-  for (uint i = 1; i < req(); ++i) {
-    res->set_req(i, in(i));
-  }
 
-  if (allocate) {
-    // Re-execute if buffering triggers deoptimization
-    PreserveReexecuteState preexecs(kit);
-    kit->jvms()->set_should_reexecute(true);
-    Node* klass_node = kit->makecon(TypeKlassPtr::make(vk));
-    Node* alloc_oop  = kit->new_instance(klass_node, nullptr, nullptr, true);
-    AllocateNode* alloc = AllocateNode::Ideal_allocation(alloc_oop);
-    alloc->_larval = true;
+  // Re-execute if buffering triggers deoptimization
+  PreserveReexecuteState preexecs(kit);
+  kit->jvms()->set_should_reexecute(true);
+  Node* klass_node = kit->makecon(TypeKlassPtr::make(vk));
+  Node* alloc_oop  = kit->new_instance(klass_node, nullptr, nullptr, true);
+  AllocateNode* alloc = AllocateNode::Ideal_allocation(alloc_oop);
+  alloc->_larval = true;
 
-    store(kit, alloc_oop, alloc_oop, vk);
-    res->set_oop(alloc_oop);
-  }
-  // TODO 8239003
-  //res->set_type(TypeInlineType::make(vk, true));
-  res = kit->gvn().transform(res)->as_InlineType();
-  assert(!allocate || res->is_allocated(&kit->gvn()), "must be allocated");
-  return res;
+  store(kit, alloc_oop, alloc_oop, vk);
+  return alloc_oop;
```

Comment:
Why not setting the oop to allocated buffer and return the `InlineTypeNode` ? We may have optimizations for Unsafe access for `InlineTypeNode` input.

Reply:
Any feedback on this please?

Reply:
The whole idea is to avoid scalarization for value objects in laval state and not construct InlineTypeNode till we encounter a finishPrivateBuffer. A value object is transitioned into larval state to make it mutable, since Unsafe.put* APIs do not modify the JVM state hence to be safe it's better to keep value objects as oops b/w make and finish private buffer, patch also disables re-materialization of InlineTypeNode from updated memory state in Unsafe.put* inline expanders.

Reply:
Yes, I understand your intention. And it makes sense to me. My question is whether we can keep the InlineType with a larval state before, and forbiting the scalarization for it between make/finish private buffer. Changing `InlineType` totally to the oop buffer looks too heavy to me.

Reply:
Generating InlineTypeNode in make_larval will also necessitate re-materialization of such a node in Unsafe.put* expander which is problematic, given that put only modifies the memory state, totally avoiding creation of InlineTypeNode for value objects in larval state will fix this.

Reply:
OK. Thanks!